text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,484 @@
+happens in a different trace. Furthermore, processing operations often are not
+covered by messaging libraries and SDKs, but take place in application code.
+Consistently creating spans for "Processing" operations would require either
+effort from the application owner to correctly instrument those operations, or
+additional capabilities of messaging libraries and SDKs (e. g. hooks for
+processing callbacks, which can then be instrumented by the libraries or SDKs).
+
+While it is possible to create "Process" spans and correlate those with
+consumer traces in certain cases, this is not something that can be generally
+required. Therefore, it is more feasible to require the creation of "Deliver"
+spans (for push-based APIs) or "Receive" spans (for pull-based APIs) to
+correlate producer with consumer traces.
+
+#### Instrumenting push-based scenarios
+
+In push-based consumer scenarios, the delivery of messages is not initiated by
+the application code. Instead, callbacks or handlers are registered and then
+called by messaging SDKs to forward messages to the application.
+
+A "Deliver" span covers the call of such a callback or handler and should link
+to the "Create" or "Publish" spans of all messages that are forwarded via the
+respective call. Depending on the use case, "Deliver" spans can correlate with
+"Process" spans or other spans modelling processing operations.
+
+#### Instrumenting pull-based scenarios
+
+In pull-based consumer scenarios, the delivery of messages is requested by the
+application code. This usually involves a blocking call, which returns zero or
+more messages on completion.
+
+A "Receive" span covers such calls and should link to the "Create" or "Publish"
+spans of all messages that are forwarded via the respective call. Depending on
+the use case, "Receive" spans can correlate with "Process" spans or other spans
+modelling processing operations.
+
+#### General considerations for both push-based and pull-based scenarios
+
+The operations modelled by "Deliver" or "Receive" spans do not strictly refer
+to receiving the message from intermediaries, but instead refer to the
+application receiving messages for processing. If messages are fetched from the
+intermediary and forwarded to the application in one go, the whole operation
+might be covered by a "Deliver" or "Receive" span. However, libraries or SDKs
+might pre-fetch messages from intermediaries and cache those messages, and only
+forward messages to the application at a later time. In this case, the
+operation of pre-fetching and caching should not be covered by the "Deliver" or
+"Receive" spans.
+
+Operations covered by "Deliver" or "Receive" can forward zero messages (e. g.
+to notify the application that no message is available for processing), one
+message, or multiple messages (a batch of messages). "Deliver" and "Receive"
+spans should link to the "Create" or "Publish" span of the messages forwarded,
```

Comment:
If I got right your doubt I think that linking makes more sense when it's not just one consumer getting the same message. The Azure SDK issue you are referring to is related to Service Bus using a queue right? So I can see just one consumer getting the produced message and having everything in one trace.
Different if using Event Hub or Kafka where different consumers (in different consumer groups) could get same produced message, so the need for a different trace linked to the producer one.
How do you get one trace in this case? The producer side should be part of different traces related to different consumers.

Reply:
The reasoning to use links exclusively for producer/consumer relationships was the fact that it is modelling an asynchronous relationship (where parent/child relationship are usually synchronous).

You could have a producer trace for publishing a single message, but the message is only consumed hours or days later. Or if there are multiple non-batch consumers, each of those will end up as part of the producer trace.

There definitely are cases where parent/child relationships are the more intuitive solution for end users. However, there are also cases where the results might be confusing.

Anyway, let's discuss it in the next workgroup meeting.

Reply:
>  Different if using Event Hub or Kafka where different consumers (in different consumer groups) could get same produced message, so the need for a different trace linked to the producer one.
> How do you get one trace in this case? The producer side should be part of different traces related to different consumers.

Each consumer would create a new span that would belong to the same trace - they are be two different branches in trace tree. I don't see any problem with it.

> You could have a producer trace for publishing a single message, but the message is only consumed hours or days later. Or if there are multiple non-batch consumers, each of those will end up as part of the producer trace.

There is no requirement for traces to be short - still, I don't see a problem here. If you'd like to optimize tooling for tail-based sampling or other scenarios like that - long traces don't really impose problems - you'd just buffer parts of these traces at different times. Note that regardless of link/parent choice here, users would create individual processing spans and would prefer using parent-child relationships. I.e. there is never a guarantee of a short trace.

Reply:
> There is no requirement for traces to be short - still, I don't see a problem here.

It's not a requirement, it's just that while some customers want both producer and consumer operations in the same trace, that doesn't mean that all (or most) customer will be happy with such a solution.

For the record, I remembered some other reasons why we initially went with links exclusively:
* Consistency over all messaging scenarios. Even if we'd recommend parent/child relationships for single-message scenarios, that might not be possible in pull-based consumer scenarios. If a user calls a `Receive` method to receive message, this method might already have a parent context (in addition to the creation context of the message).
* Keeping possibilities open for intermediary instrumentation. If there's a parent/child relationship between producer and consumer spans, it's not clear how intermediary instrumentation can fit into the picture. It's not impossible, but I think it will be more cumbersome.

However, in my opinion instrumentors should still be allowed to use parent/child relationships for single-message scenarios when it's possible and when it makes sense for them, without breaking compatibility with the semantic conventions (we're not using any `MUST` statements currently). It's not clear to me though whether that's something we should recommend.

I put this up as a discussion point for the next workgroup meeting, I'd love to hear other people's thoughts.

Reply:
This was discussed in today's workgroup meeting.

There's consensus that for some single-message scenarios users prefer parent/child relationships between producer and consumer spans instead of links.  However, using parent/child relationships makes it impossible to make a consumer span a child of an ambient context, instead it needs to be a service root span.

We agreed on strongly recommending using links for connecting producer and consumer spans. However, a parent/child relationship may be used in addition to the link in scenarios where this makes sense. With this solution, backends can consistently rely on links, and instrumentations can provide customers with single-trace solutions for scenarios where it makes sense.

Reply:
I added this additional text and examples to the OTEP.

Reply:
> We agreed on strongly recommending using links for connecting producer and consumer spans.

I don't agree with this.  I think the parent/child relationship should be associated with whichever has the strongest affinity. In some cases it might be the producer span in other cases the ambient span. As such, I like the proposal to use parent/child when just a single message is consumed.  I would like to extend that proposal slightly:

Parent/child relationship should be used with the remote span instead of the ambient span if the following are true:
-  the consumed batch contains a single message OR all messages share the same trace id OR there is no ambient span OR a setting is configured for the instrumentation to prefer this behavior.
- (if possible to determine) the message was created/produced within a configurable threshold.

In either case, the span representing the receive/deliver/handle of an individual message should be put in scope for the duration that the message is being processed to signify the handling of that message.

The reason to use a configurable timeout as an additional signal for determining the relationship is that these situations likely represent message reprocessing and much less strongly coupled to the original producer.  In many tracing systems, gaps of long periods of time also are not handled well and visually distort the trace graph. I understand that creation date is not always available to all instrumentation which is why I added "if possible".